File: graphlib-2/ts/src/components.ts
// Cycle detection, topological sort, connected components, strongly connected components

import {
  HasCycleResult,
  IsDAGResult,
  TopologicalSortResult,
  ConnectedComponentsResult,
  IsConnectedResult,
} from './types';
import { getGraph } from './graph';

export function hasCycle(graphId: string): HasCycleResult {
  const graph = getGraph(graphId);
  if (!graph || graph.nodes.size === 0) {
    return { has_cycle: false, cycle: [] };
  }

  if (graph.directed) {
    return hasCycleDirected(graph);
  } else {
    return hasCycleUndirected(graph);
  }
}

function hasCycleDirected(
  graph: NonNullable<ReturnType<typeof getGraph>>
): HasCycleResult {
  const WHITE = 0; // unvisited
  const GRAY = 1; // in progress
  const BLACK = 2; // finished

  const color = new Map<string, number>();
  const parent = new Map<string, string>();

  for (const node of graph.nodes) {
    color.set(node, WHITE);
  }

  function dfs(node: string): string[] | null {
    color.set(node, GRAY);

    const neighbors = Array.from(graph.adjacency.get(node)?.keys() ?? []).sort();
    for (const neighbor of neighbors) {
      if (color.get(neighbor) === GRAY) {
        // Found back edge - cycle detected
        const cycle: string[] = [neighbor];
        let current = node;
        while (current !== neighbor) {
          cycle.unshift(current);
          current = parent.get(current) ?? neighbor;
        }
        cycle.unshift(neighbor);
        return cycle;
      }
      if (color.get(neighbor) === WHITE) {
        parent.set(neighbor, node);
        const result = dfs(neighbor);
        if (result) return result;
      }
    }

    color.set(node, BLACK);
    return null;
  }

  for (const node of Array.from(graph.nodes).sort()) {
    if (color.get(node) === WHITE) {
      const cycle = dfs(node);
      if (cycle) {
        return { has_cycle: true, cycle };
      }
    }
  }

  return { has_cycle: false, cycle: [] };
}

function hasCycleUndirected(
  graph: NonNullable<ReturnType<typeof getGraph>>
): HasCycleResult {
  const visited = new Set<string>();
  const parent = new Map<string, string | null>();

  function dfs(node: string, parentNode: string | null): string[] | null {
    visited.add(node);
    parent.set(node, parentNode);

    const neighbors = Array.from(graph.adjacency.get(node)?.keys() ?? []).sort();
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor)) {
        const result = dfs(neighbor, node);
        if (result) return result;
      } else if (neighbor !== parentNode) {
        // Found cycle - reconstruct it
        const cycle: string[] = [neighbor];
        let current = node;
        while (current !== neighbor) {
          cycle.unshift(current);
          current = parent.get(current) ?? neighbor;
        }
        cycle.unshift(neighbor);
        return cycle;
      }
    }

    return null;
  }

  for (const node of Array.from(graph.nodes).sort()) {
    if (!visited.has(node)) {
      const cycle = dfs(node, null);
      if (cycle) {
        return { has_cycle: true, cycle };
      }
    }
  }

  return { has_cycle: false, cycle: [] };
}

export function isDAG(graphId: string): IsDAGResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { is_dag: false };
  }

  // DAG is only defined for directed graphs
  if (!graph.directed) {
    return { is_dag: false };
  }

  // Empty directed graph is a DAG
  if (graph.nodes.size === 0) {
    return { is_dag: true };
  }

  const cycleResult = hasCycle(graphId);
  return { is_dag: !cycleResult.has_cycle };
}

export function topologicalSort(graphId: string): TopologicalSortResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }

  // Only works for directed acyclic graphs
  if (!graph.directed) {
    return { success: false, error: 'not_a_dag' };
  }

  // Check for cycles
  const cycleResult = hasCycle(graphId);
  if (cycleResult.has_cycle) {
    return { success: false, error: 'not_a_dag' };
  }

  // Kahn's algorithm
  const inDegree = new Map<string, number>();
  for (const node of graph.nodes) {
    inDegree.set(node, 0);
  }

  for (const [, neighbors] of graph.adjacency) {
    for (const neighbor of neighbors.keys()) {
      inDegree.set(neighbor, (inDegree.get(neighbor) ?? 0) + 1);
    }
  }

  const queue: string[] = [];
  for (const [node, degree] of inDegree) {
    if (degree === 0) {
      queue.push(node);
    }
  }
  queue.sort();

  const order: string[] = [];

  while (queue.length > 0) {
    queue.sort(); // Ensure deterministic ordering
    const current = queue.shift()!;
    order.push(current);

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        const newDegree = inDegree.get(neighbor)! - 1;
        inDegree.set(neighbor, newDegree);
        if (newDegree === 0) {
          queue.push(neighbor);
        }
      }
    }
  }

  return { success: true, order };
}

export function connectedComponents(graphId: string): ConnectedComponentsResult {
  const graph = getGraph(graphId);
  if (!graph || graph.nodes.size === 0) {
    return { count: 0, components: [] };
  }

  const visited = new Set<string>();
  const components: string[][] = [];

  function bfs(startNode: string): string[] {
    const component: string[] = [];
    const queue: string[] = [startNode];
    visited.add(startNode);

    while (queue.length > 0) {
      const current = queue.shift()!;
      component.push(current);

      // For connectivity, we need to traverse both directions for directed graphs
      const outNeighbors = graph!.adjacency.get(current) ?? new Map();
      const inNeighbors = graph!.inEdges.get(current) ?? new Map();

      const allNeighbors = new Set([...outNeighbors.keys(), ...inNeighbors.keys()]);

      for (const neighbor of allNeighbors) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push(neighbor);
        }
      }
    }

    return component.sort();
  }

  for (const node of Array.from(graph.nodes).sort()) {
    if (!visited.has(node)) {
      const component = bfs(node);
      components.push(component);
    }
  }

  return { count: components.length, components };
}

export function stronglyConnectedComponents(
  graphId: string
): ConnectedComponentsResult {
  const graph = getGraph(graphId);
  if (!graph || graph.nodes.size === 0) {
    return { count: 0, components: [] };
  }

  // For undirected graphs, SCC is same as CC
  if (!graph.directed) {
    return connectedComponents(graphId);
  }

  // Kosaraju's algorithm
  const visited = new Set<string>();
  const finishOrder: string[] = [];

  // First DFS to get finish order
  function dfs1(node: string): void {
    visited.add(node);
    const neighbors = Array.from(graph!.adjacency.get(node)?.keys() ?? []).sort();
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor)) {
        dfs1(neighbor);
      }
    }
    finishOrder.push(node);
  }

  for (const node of Array.from(graph.nodes).sort()) {
    if (!visited.has(node)) {
      dfs1(node);
    }
  }

  // Build transpose graph edges (use inEdges)
  visited.clear();
  const components: string[][] = [];

  // Second DFS on transpose in reverse finish order
  function dfs2(node: string, component: string[]): void {
    visited.add(node);
    component.push(node);
    const neighbors = Array.from(graph!.inEdges.get(node)?.keys() ?? []).sort();
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor)) {
        dfs2(neighbor, component);
      }
    }
  }

  for (let i = finishOrder.length - 1; i >= 0; i--) {
    const node = finishOrder[i];
    if (!visited.has(node)) {
      const component: string[] = [];
      dfs2(node, component);
      components.push(component.sort());
    }
  }

  return { count: components.length, components };
}

export function isConnected(graphId: string): IsConnectedResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { is_connected: true };
  }

  if (graph.nodes.size === 0) {
    return { is_connected: true };
  }

  if (graph.nodes.size === 1) {
    return { is_connected: true };
  }

  // For directed graphs, check weak connectivity
  const ccResult = connectedComponents(graphId);
  return { is_connected: ccResult.count === 1 };
}

File: graphlib-2/go/lib/components.go
package lib

import (
	"sort"
)

// HasCycle checks if the graph contains any cycle
func (g *Graph) HasCycle() (bool, []string) {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if len(g.Nodes) == 0 {
		return false, []string{}
	}

	if g.Directed {
		return g.hasCycleDirected()
	}
	return g.hasCycleUndirected()
}

func (g *Graph) hasCycleDirected() (bool, []string) {
	// Colors: 0 = white (unvisited), 1 = gray (in progress), 2 = black (done)
	color := make(map[string]int)
	parent := make(map[string]string)
	var cycleStart, cycleEnd string

	var dfs func(node string) bool
	dfs = func(node string) bool {
		color[node] = 1 // gray

		neighbors := make([]string, 0)
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if color[neighbor] == 1 {
				// Back edge found - cycle detected
				cycleStart = neighbor
				cycleEnd = node
				return true
			}
			if color[neighbor] == 0 {
				parent[neighbor] = node
				if dfs(neighbor) {
					return true
				}
			}
		}

		color[node] = 2 // black
		return false
	}

	// Get nodes in sorted order for deterministic behavior
	nodes := make([]string, 0, len(g.Nodes))
	for n := range g.Nodes {
		nodes = append(nodes, n)
	}
	sort.Strings(nodes)

	for _, node := range nodes {
		if color[node] == 0 {
			if dfs(node) {
				// Reconstruct cycle
				cycle := []string{}
				current := cycleEnd
				for current != cycleStart {
					cycle = append([]string{current}, cycle...)
					current = parent[current]
				}
				cycle = append([]string{cycleStart}, cycle...)
				cycle = append(cycle, cycleStart)
				return true, cycle
			}
		}
	}

	return false, []string{}
}

func (g *Graph) hasCycleUndirected() (bool, []string) {
	visited := make(map[string]bool)
	parent := make(map[string]string)
	var cycleStart, cycleEnd string

	var dfs func(node, par string) bool
	dfs = func(node, par string) bool {
		visited[node] = true

		neighbors := make([]string, 0)
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				parent[neighbor] = node
				if dfs(neighbor, node) {
					return true
				}
			} else if neighbor != par {
				// Cycle found
				cycleStart = neighbor
				cycleEnd = node
				return true
			}
		}
		return false
	}

	// Check for self-loops first
	for node := range g.Nodes {
		if _, exists := g.Adj[node][node]; exists {
			return true, []string{node, node}
		}
	}

	nodes := make([]string, 0, len(g.Nodes))
	for n := range g.Nodes {
		nodes = append(nodes, n)
	}
	sort.Strings(nodes)

	for _, node := range nodes {
		if !visited[node] {
			if dfs(node, "") {
				// Reconstruct cycle
				cycle := []string{}
				current := cycleEnd
				for current != cycleStart {
					cycle = append([]string{current}, cycle...)
					current = parent[current]
				}
				cycle = append([]string{cycleStart}, cycle...)
				cycle = append(cycle, cycleStart)
				return true, cycle
			}
		}
	}

	return false, []string{}
}

// IsDAG checks if the graph is a Directed Acyclic Graph
func (g *Graph) IsDAG() bool {
	g.mu.RLock()
	directed := g.Directed
	g.mu.RUnlock()

	if !directed {
		return false
	}

	hasCycle, _ := g.HasCycle()
	return !hasCycle
}

// TopologicalSort returns a topological ordering of the graph
func (g *Graph) TopologicalSort() (bool, []string, string) {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if !g.Directed {
		return false, nil, "not_a_dag"
	}

	// Check for cycles
	hasCycle, _ := g.hasCycleDirected()
	if hasCycle {
		return false, nil, "not_a_dag"
	}

	// Kahn's algorithm
	inDegree := make(map[string]int)
	for node := range g.Nodes {
		inDegree[node] = 0
	}

	for _, edges := range g.Adj {
		for to := range edges {
			inDegree[to]++
		}
	}

	// Queue of nodes with no incoming edges
	queue := make([]string, 0)
	for node, deg := range inDegree {
		if deg == 0 {
			queue = append(queue, node)
		}
	}
	sort.Strings(queue)

	result := make([]string, 0, len(g.Nodes))

	for len(queue) > 0 {
		// Sort queue to ensure deterministic order
		sort.Strings(queue)
		node := queue[0]
		queue = queue[1:]
		result = append(result, node)

		neighbors := make([]string, 0)
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	if len(result) != len(g.Nodes) {
		return false, nil, "not_a_dag"
	}

	return true, result, ""
}

// ConnectedComponents finds all connected components
func (g *Graph) ConnectedComponents() [][]string {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if len(g.Nodes) == 0 {
		return [][]string{}
	}

	visited := make(map[string]bool)
	components := make([][]string, 0)

	// For directed graphs, we compute weakly connected components
	// by treating the graph as undirected
	adjUndirected := make(map[string]map[string]bool)
	for node := range g.Nodes {
		adjUndirected[node] = make(map[string]bool)
	}

	for from, edges := range g.Adj {
		for to := range edges {
			adjUndirected[from][to] = true
			adjUndirected[to][from] = true
		}
	}

	var bfs func(start string) []string
	bfs = func(start string) []string {
		component := make([]string, 0)
		queue := []string{start}
		visited[start] = true

		for len(queue) > 0 {
			node := queue[0]
			queue = queue[1:]
			component = append(component, node)

			neighbors := make([]string, 0)
			for neighbor := range adjUndirected[node] {
				neighbors = append(neighbors, neighbor)
			}
			sort.Strings(neighbors)

			for _, neighbor := range neighbors {
				if !visited[neighbor] {
					visited[neighbor] = true
					queue = append(queue, neighbor)
				}
			}
		}

		sort.Strings(component)
		return component
	}

	nodes := make([]string, 0, len(g.Nodes))
	for n := range g.Nodes {
		nodes = append(nodes, n)
	}
	sort.Strings(nodes)

	for _, node := range nodes {
		if !visited[node] {
			component := bfs(node)
			components = append(components, component)
		}
	}

	// Sort components by first element
	sort.Slice(components, func(i, j int) bool {
		if len(components[i]) == 0 {
			return true
		}
		if len(components[j]) == 0 {
			return false
		}
		return components[i][0] < components[j][0]
	})

	return components
}

// StronglyConnectedComponents finds SCCs using Kosaraju's algorithm
func (g *Graph) StronglyConnectedComponents() [][]string {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if len(g.Nodes) == 0 {
		return [][]string{}
	}

	// For undirected graphs, return same as connected components
	if !g.Directed {
		g.mu.RUnlock()
		components := g.ConnectedComponents()
		g.mu.RLock()
		return components
	}

	// Kosaraju's algorithm
	// Step 1: DFS to get finish order
	visited := make(map[string]bool)
	finishOrder := make([]string, 0, len(g.Nodes))

	var dfs1 func(node string)
	dfs1 = func(node string) {
		visited[node] = true

		neighbors := make([]string, 0)
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				dfs1(neighbor)
			}
		}
		finishOrder = append(finishOrder, node)
	}

	nodes := make([]string, 0, len(g.Nodes))
	for n := range g.Nodes {
		nodes = append(nodes, n)
	}
	sort.Strings(nodes)

	for _, node := range nodes {
		if !visited[node] {
			dfs1(node)
		}
	}

	// Step 2: Build transpose graph
	transpose := make(map[string]map[string]bool)
	for node := range g.Nodes {
		transpose[node] = make(map[string]bool)
	}
	for from, edges := range g.Adj {
		for to := range edges {
			transpose[to][from] = true
		}
	}

	// Step 3: DFS on transpose in reverse finish order
	visited = make(map[string]bool)
	components := make([][]string, 0)

	var dfs2 func(node string, component *[]string)
	dfs2 = func(node string, component *[]string) {
		visited[node] = true
		*component = append(*component, node)

		neighbors := make([]string, 0)
		for neighbor := range transpose[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				dfs2(neighbor, component)
			}
		}
	}

	// Process in reverse finish order
	for i := len(finishOrder) - 1; i >= 0; i-- {
		node := finishOrder[i]
		if !visited[node] {
			component := make([]string, 0)
			dfs2(node, &component)
			sort.Strings(component)
			components = append(components, component)
		}
	}

	// Sort components by first element
	sort.Slice(components, func(i, j int) bool {
		if len(components[i]) == 0 {
			return true
		}
		if len(components[j]) == 0 {
			return false
		}
		return components[i][0] < components[j][0]
	})

	return components
}

// IsConnected checks if the graph is connected
func (g *Graph) IsConnected() bool {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if len(g.Nodes) == 0 {
		return true
	}

	// Get any starting node
	var start string
	for n := range g.Nodes {
		start = n
		break
	}

	// BFS treating graph as undirected (for weak connectivity in directed graphs)
	visited := make(map[string]bool)
	queue := []string{start}
	visited[start] = true

	// Build undirected adjacency for directed graphs
	adjUndirected := make(map[string]map[string]bool)
	for node := range g.Nodes {
		adjUndirected[node] = make(map[string]bool)
	}
	for from, edges := range g.Adj {
		for to := range edges {
			adjUndirected[from][to] = true
			adjUndirected[to][from] = true
		}
	}

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		for neighbor := range adjUndirected[node] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return len(visited) == len(g.Nodes)
}
